### PR TITLE
[`GPTNeoX`] Faster rotary embedding for GPTNeoX (based on llama changes)

### LINKED ISSUE #25813: GPTNeoXRotaryEmbedding has a defect when using sin/cos cache
### System Info

- `transformers` version: 4.31.0
- Platform: Linux-5.10.112-005.ali5000.alios7.x86_64-x86_64-with-glibc2.17
- Python version: 3.8.13
- Huggingface_hub version: 0.12.1
- PyTorch version (GPU?): 1.13.1+cu117 (False)
- Tensorflow version (GPU?): 2.11.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

Text models:@ArthurZucker  and @younesbelkada

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

in transformers/models/gpt_neox/modeling_gpt_neox.py, line320:
```python
return self.cos_cached[:seq_len, ...].to(x.device), self.sin_cached[:seq_len, ...].to(x.device)
```
lack 2 dimensions before seq_lenth.
This will not cause bug, because seq_lenth is always larger than 1, but it will cause a defect use of cache by use the whole cache ever, maybe lead to poor performance when inference?

### Expected behavior

Maybe the right code should be:
```python
return self.cos_cached[:, :, :seq_len, ...].to(x.device), self.sin_cached[:, :, :seq_len, ...].to(x.device)
```

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Fixes #25813 which indicates that ROPE is slow. It's a follow up of #22785, were ROPE was improved for Llama model.

### CODE REVIEW

File: src/transformers/models/idefics/modeling_idefics.py [hunk 1]
```diff
@@ -494,33 +494,37 @@ def forward(self, hidden_states):
...
+        self.max_seq_len_cached = seq_len
+        t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)
+
         freqs = torch.einsum("i,j->ij", t, self.inv_freq)
         # Different from paper, but it uses a different permutation in order to obtain the same calculation
         emb = torch.cat((freqs, freqs), dim=-1)
-        self.register_buffer("cos_cached", emb.cos()[None, None, :, :], persistent=False)
-        self.register_buffer("sin_cached", emb.sin()[None, None, :, :], persistent=False)
+        self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(dtype), persistent=False)
+        self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(dtype), persistent=False)
```
Comment (amyeroberts) [inline src/transformers/models/idefics/modeling_idefics.py:521]: Out of interest, why do we register a buffer in IDEFICS and just set attributes in the other models?

File: src/transformers/models/idefics/modeling_idefics.py [hunk 2]
```diff
@@ -494,33 +494,37 @@ def forward(self, hidden_states):
...
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2).float().to(device) / self.dim))
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
 
         # Build here to make `torch.jit.trace` work.
-        self.max_seq_len_cached = max_position_embeddings
-        t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
+        self._set_cos_sin_cache(
+            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
```
Comment (amyeroberts) [inline src/transformers/models/idefics/modeling_idefics.py:510]: This is a potential breaking change, as the dtype is no longer necessarily the same

File: src/transformers/models/idefics/modeling_idefics.py [hunk 3]
```diff
@@ -494,33 +494,37 @@ def forward(self, hidden_states):
...
-        # This `if` block is unlikely to be run after we build sin/cos in `__init__`. Keep the logic here just in case.
         if seq_len > self.max_seq_len_cached:
-            self.max_seq_len_cached = seq_len
-            t = torch.arange(self.max_seq_len_cached, device=x.device, dtype=self.inv_freq.dtype)
-            freqs = torch.einsum("i,j->ij", t, self.inv_freq)
-            # Different from paper, but it uses a different permutation in order to obtain the same calculation
-            emb = torch.cat((freqs, freqs), dim=-1).to(x.device)
-            self.register_buffer("cos_cached", emb.cos()[None, None, :, :], persistent=False)
-            self.register_buffer("sin_cached", emb.sin()[None, None, :, :], persistent=False)
+            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
```
Comment (amyeroberts) [inline src/transformers/models/idefics/modeling_idefics.py:526]: Same here about the dtype

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Only question is about the dtype casting in Idefics. Could you run slow tests and some checks on the effects of the model outputs when rope scaling is used?

File: src/transformers/models/idefics/modeling_idefics.py (hunk 1 above)
Comment (gante) [reply src/transformers/models/idefics/modeling_idefics.py:521]: @amyeroberts I had the same question reading your comment, and decided to dig deeper! 

According to [this issue](https://github.com/pytorch/pytorch/issues/18056), it's the only way to a) keep it out of `state_dict` (non-persistent buffers and variables fit this requirement) while b) preserving the ability to be impacted by `to(<type/device>)` at a module-level (buffers and parameters fit this requirement)!

(the whole issue is an interesting read)

**Review by gante — APPROVED** [review]
LGTM (and nice touch adding the copied from!)

File: src/transformers/models/idefics/modeling_idefics.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/idefics/modeling_idefics.py:510]: It's not as the dtype of ` inv_freq`  if not specified will always be the same as ` torch.get_default_dtype`

**Comment by LysandreJik** [comment]
cc @StellaAthena FYI, this PR should greatly speed up the ROPE embeddings of the GPTNeoX model, similarly to how it was done for the LLaMa model.

Let us know if you want to review/have any comments!

**Review by LysandreJik — APPROVED** [review]

**Comment by ArthurZucker** [comment]
cc @Narsil as this touches buffers that will no longer be persistent, will wait for you in case this is conflicting with TGI?

**Pull request merged**